### PR TITLE
refactor(es/typescript): Reimplement TypeScript pass

### LINKED ISSUE #5197: Namespace with `declare`d members should not be type only
### Describe the bug

A namespace with only `declare`d values is treated as type-only and elided at transpilation, it should be transpiled to an empty namespace.

### Input code

```typescript
export namespace foo {
  export declare const bar: number;
}
```

### Config

_No response_

### Playground link

https://play.swc.rs/?version=1.2.212&code=H4sIAAAAAAAAA0utKMgvKlHIS8xNLS5ITE5VSMvPV6jmUlBIhUikpCbnJBalKiTn5xWXKCQlFlkp5JXmJqUWWXPVAgDtD19xPAAAAA%3D%3D&config=H4sIAAAAAAAAA0WOSwrDMAxE76J1Fm0XhfoOPYRwleDgH5ICNcZ3r21SshtmRk9TYRcLpkJGFuKhpETFLxjQkkksu6ywgEq3lA9qXSNvpL1B8rjdXz31KQmBWdELLRBcdGsZLJtCZhK5Ioyb%2FzdbZ4X0OYZR57vJfEK7GOedk%2FdZnBt%2BJYyKhrcAAAA%3D

### Expected behavior

TSC output:
```js
export var foo;
(function (foo) {
})(foo || (foo = {}));
```
https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAOwIYFtgGcxIMbDgMwgjgG8AoOOUSWOAE2GwBsko9sIF14AjVgLkQBXFD2BQA3GQC+ZIA

### Actual behavior

Produces empty module.

### Version

swc_ecma_parser 0.111.0, swc_ecma_transforms_typescript 0.137.0

### Additional context

_No response_

### LINKED ISSUE #5259: Exported mutable variable in namespace with setter function does not mutate value
### Describe the bug

If an export from a namespace is mutable, and a function in the namespace mutates the value, the exported value is not mutated.

### Input code

```typescript
namespace A {
    export let a = 5;

    export function setA(value: number) {
        a = value;
    }
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": false
    },
    "target": "es5",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "commonjs"
  },
  "minify": false,
  "isModule": true
}
```

### Playground link

https://play.swc.rs/?version=1.2.218&code=H4sIAAAAAAAAA8tLzE0tLkhMTlVwVKjmUgCC1IqC%2FKIShZzUEoVEBVsFU2suZOG00rzkksz8PIXi1BJHjbLEnNJUK4W80tyk1CJNqAEgANIJlrQGC9Vy1QIA8jab8GkAAAA%3D&config=H4sIAAAAAAAAA0WNSwrDMAxE76J1tt3kDj2ESZXg4B8aBWqM7165OGQnzedNoxMbrY2KE7CMCzWp%2B9JKWgtjE1%2BUFlKYtLsA7vY4OVgtwniZF3IGT3eh6JPf6yBtORZh4LFcOsKd7AaK%2BXMNof3HDGiVmNMJ6g9olj3eM61ycf8BnwbyVboAAAA%3D

### Expected behavior

`tsc` output correctly mutates the value:

```js
"use strict";
var A;
(function (A) {
    A.a = 5;
    function setA(value) {
        A.a = value;
    }
    A.setA = setA;
})(A || (A = {}));
```

### Actual behavior

Mutates the variable internal to the namespace but not the exported value from the namespace.

```js
"use strict";
var A;
(function(A) {
    var setA = function setA(value) {
        a = value;
    };
    var a = A.a = 5;
    A.setA = setA;
})(A || (A = {}));

```

### Version

1.2.218

### Additional context

_No response_

### LINKED ISSUE #7177: Compilation omits import aliases for TypeScript
### Describe the bug

`swc` omits  import aliases
```typescript
import * as NS from './a';

import Alias = NS;
import A = Alias.A;

export { A };
```
what results in the runtime error
```
var Alias = NS;
            ^
ReferenceError: NS is not defined
```

### Input code

```typescript
import * as NS from './a';

import Alias = NS;
import A = Alias.A;

export { A };
```

### Config

```json
{
  "module": {
    "type": "commonjs",
    "strict": true
  },
  "jsc": {
    "target": "es2022",
    "parser": {
      "syntax": "typescript"
    }
  }
}
```

### Playground link

https://play.swc.rs/?version=1.3.43&code=H4sIAAAAAAAAA8vMLcgvKlHQUkgsVvALVkgrys9VUNfTT1S35uXi5cqEyDrmZAKlbYEKrBFiQD5YXM8RrDS1AixcDZSoBQoAAJhh2QdYAAAA&config=H4sIAAAAAAAAA1VPSQ7DIAy88wrkcw8Vx%2F6hj0DUiYjYhB2pKMrfCymkzc2exTPehJSwkIGH3OpYl6QzYT73ilAJrN8VAS4JyWSbGG6DZWrUpB3hAe1fBljnGbm5kNRdqe4AFyPhcHTM22Cn8p9pok8Zia7CJtVhdnhNFD0VfHytB9l%2FaX1bg3rNx7AQ%2FJQj8bwOlp7DznlFsX8AqE9BlhwBAAA%3D

### Expected behavior

`NS` variable should be imported and declared

### Actual behavior

`NS` is not defined

### Version

1.3.42

### Additional context

@kdy1 please let me know if it is clear

### LINKED ISSUE #7453: References to destructured exports in namespaces aren't resolvable at runtime
### Describe the bug

Inside a namespace, referring to an exported property in the namespace generally works because it gets emitted as a `var` binding. However, when the export is defined using destructuring syntax, there are no such bindings and so any references to exports declared that way fail at runtime.

### Input code

```typescript
export namespace cli {
  export const { options, args } = parseCommandLine();
  export function run() {
    const command = args.shift();
  }
}
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": false,
      "decorators": true
    },
    "target": "es2022",
    "loose": false,
    "transform": {
      "legacyDecorator": true,
      "decoratorMetadata": true
    },
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "commonjs"
  },
  "minify": false,
  "isModule": true
}
```

### Playground link

https://play.swc.rs/?version=1.3.60&code=H4sIAAAAAAAAA0XMQQqDMBCF4b2neEuF0guUrtx6iWE62oCZhEwEQXL3xkZw9%2BDxf7LHkDKUvFgkFvDqcHSAtIODWsaBELOr8wFKi6HgjUjJZAzek34mp9IPrzubN%2BUzQNq0H%2F4gLotbUoWTetrXzbm1pSs%2FQPhzc48AAAA%3D&config=H4sIAAAAAAAAA22Quw6DMAxF934F8tyhYuzclY%2BwgkFB5CHbSEWIf294hFLRLfZ1zr32dCsK6MTAs5jSMxURWYiPOnVk9Irv1AEdI4lhGxXuWVVZpAZ7oaNXkwmMGliSpDzQKsybDorcki48kvJRljsL%2BhCEflmgjF6awO4cqKcWzfjKJrvH1b0ixRoV%2F4Vw1ttmPFNNcJFJ5LKNQ9%2F2R7CNcttJ4EI9rOJ%2BvuVEy2qJ5oLvBL6T2fGgg5Uqf18Dzh87elJPjwEAAA%3D%3D

### Expected behavior

Either emit a `var` binding for each destructured property, or rewrite the reference to the exported property from `args` to `cli.args` (`tsc` does the latter).

### Actual behavior

```js
var cli;
(function(cli) {
    ({ options: cli.options , args: cli.args  } = parseCommandLine());
    function run() {
        const command = args.shift();
    }
    cli.run = run;
})(cli || (cli = {}));
```

At runtime, `ReferenceError: args is not defined`

### Version

1.3.60

### Additional context

Here's the [tsc output](https://www.typescriptlang.org/play?module=1#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwAdYBnEAYRwFtKpVgAZLVEACgEoAueAb3hwOx5iXWgE8ANPFgBzYfGIYYTaQG0AuvAC+AbgBQIAB4EcMDPFRRKIYkTAJIWHroCQh46fhghZ3v0GpiSRliLXgAXkIScioaOkZmdj1XIxMzFHR-eBg0didnZy8Asy9qWmBwqRhZADpiAAssRAxE3XgtXU0gA), for comparison.

```js
var cli;
(function (cli) {
    var _a;
    _a = parseCommandLine(), cli.options = _a.options, cli.args = _a.args;
    function run() {
        const command = cli.args.shift();
    }
    cli.run = run;
})(cli = exports.cli || (exports.cli = {}));
```

I had a quick look but only found #3073 as a possible related issue, which was quite a while ago.

### LINKED ISSUE #7676: TS namespace containing just `declare class` incorrectly removed
### Describe the bug

TS doesn't remove namespaces that contain `declare class`, while SWC does

### Input code

```typescript
namespace NS {
  declare class D {}
}

console.log(typeof NS)
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "typescript",
      "tsx": false
    },
    "target": "es2019",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true
}
```

### Playground link

https://play.swc.rs/?version=1.3.70&code=H4sIAAAAAAAAA8tLzE0tLkhMTlXwC1ao5lJQSElNzkksSlUAksXFCi4K1bVctVxcyfl5xfk5qXo5%2BekaJZUFqflpQPWaXADOIbDEPgAAAA%3D%3D&config=H4sIAAAAAAAAA1VPSQ7CMAy85xWRzxyAAxL8gUdEwa2Csil2JaKqfydpk0Jv9iye8SykhDdpeMi5jGWJKhGmfS8IZc%2FqUxDgHJF0MpHh1FmmSg3KEq7QsjHAKo3I1YV0PV%2FuzQE2BMLuaJgz3gz5P1MHFxMSHYVVqvxo8ZgoWiq48JpWsv1S%2B24NbvAT9bD9MBh6dienCcXyBaIK9Y0XAQAA

### Expected behavior

```js
var NS;
(function (NS) {
})(NS || (NS = {}));
console.log(typeof NS);
```

### Actual behavior

```js
console.log(typeof NS);
```
and at runtime it throws that `NS` is not declared

### Version

1.3.70

### Additional context

_No response_

### LINKED ISSUE #7681: Wrong ts enum transform result
### Describe the bug

`Foo["a"]` or `Foo.a` cannot be resolved by swc.

### Input code

```typescript
enum Foo {
    "a",
    b = 1,
    c = Foo["a"],
    "d" = Foo.a,
    "e",
    "f"
}
```

### Config

```json
{
  "jsc": {
    "experimental": {
      "keepImportAssertions": true
    },
    "parser": {
      "syntax": "typescript",
      "tsx": true
    },
    "target": "es2022"
  },
  "module": {
    "type": "es6"
  },
  "isModule": false
}
```

### Playground link

[playground](https://play.swc.rs/?version=1.3.70&code=H4sIAAAAAAAAA0vNK81VcMvPV6jmUgACpUQlHTAjScFWwRDCTAYygSqigXKxEBGlFCWImF4iVCAVqk0pTYmrFgBWk1WQVAAAAA%3D%3D&config=H4sIAAAAAAAAA22OMQ7CMAxF95wi8twBZWBgY2TgEFFxUSBpItuVUlW9Oy6llIEt%2Be%2F9L0%2FGWnhwCyc76VM%2FWAtSSNiLj99U8ydiuaSSSc7MSBJyz8qFBnwrc7P2iyfFv00edatqAjIW5JZCEWg2Klz%2FzYinO8pSQnYH58B8GKR8GyLuBy%2Bjq3fcpcDXTet8ZDSzeQER25%2FK6wAAAA%3D%3D)

### Expected behavior

https://www.typescriptlang.org/play?#code/KYOwrgtgBAYg9nKBvAUFdUBEBDTAaNDAIygF4oBGAjKAYzNgQG0dMBdajTAE0wfjgA6bJ3SZg+QmIBmmFAF8gA

```JavaScript 
var Foo;
(function (Foo) {
    Foo[Foo["a"] = 0] = "a";
    Foo[Foo["b"] = 1] = "b";
    Foo[Foo["c"] = 0] = "c";
    Foo[Foo["d"] = 0] = "d";
    Foo[Foo["e"] = 1] = "e";
    Foo[Foo["f"] = 2] = "f";
})(Foo || (Foo = {}));
```

### Actual behavior

```JavaScript 
var Foo;
(function(Foo) {
    Foo[Foo["a"] = 0] = "a";
    Foo[Foo["b"] = 1] = "b";
    Foo[Foo["c"] = Foo["a"]] = "c";
    Foo[Foo["d"] = Foo.a] = "d";
    Foo[Foo["e"] = 2] = "e";
    Foo[Foo["f"] = 3] = "f";
})(Foo || (Foo = {}));
```

### Version

1.3.70

### Additional context

_No response_

### LINKED ISSUE #7791: non-null assertion in typescript changes generated javascript
### Describe the bug

Typescript documentation explicitly states that using the non-null assertion (`!`) does not change the emitted code.

But, using `swc`, this isn't the case. The following create different outputs:

`a?.b!.c ?? 0`
vs
`a?.b.c ?? 0`

At least, when targeting ES5.

The first case will generate something like:

`(_a = (a === null || a === void 0 ? void 0 : a.b).x) !== null && _a !== void 0 ? _a : 0`

while the second generates

`(_a = a  === null || a === void 0 ? void 0 : a.b.x) !== null && _a !== void 0 ? _a : 0`

Using `tsc`, these expressions generate the same output.

Works as expected in versions < 1.3.63

### Input code

```typescript
a?.b!.c ?? 0
```

### Config

```json
n/a
```

### Playground link

https://play.swc.rs/?version=1.3.75&code=H4sIAAAAAAAAA0u010tS1EtWsLdXMAAAPu9sgQwAAAA%3D&config=H4sIAAAAAAAAA1WPSw7CMAxE9z1F5DVbWHAHDmEFtwrKT7GRiKrenaRJCt3ZM2M%2Fe52UghdruKu1lKWJmJjS0ReFsxf8FAUkR2KdTBS4DFe4WjNapl3amgOCaSGpU8TXHgcbAtOId80Zb%2Bb8D9TBxUTM52CNol8snXFTR4ILz%2Fdu9kfqsQ1%2Fg19owI7FYPgxJtva7QvTya0RFQEAAA%3D%3D

### Expected behavior

`a?.b.c` runs identically to `a?.b!.c`

output
`(_a = a  === null || a === void 0 ? void 0 : a.b.x) !== null && _a !== void 0 ? _a : 0`

### Actual behavior

Using `a?.b!.c` throws an error at runtime when a is undefined, while `a?.b.c` does not.

outputs
`(_a = (a === null || a === void 0 ? void 0 : a.b).x) !== null && _a !== void 0 ? _a : 0`

### Version

1.3.63 - 1.3.75

### Additional context

_No response_

### LINKED ISSUE #7961: Cannot trim const enum
### Describe the feature

The compilation result of `tsc` is cleaner and neater.

swc
image

tsc :
image

### Babel plugin or link to the feature description

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by magic-akari**
**Description:**

## Bugfixes
- Exported `let`/`var` declarations in TypeScript namespaces should be mutable.
- Fix missing declaration of complex exported patterns in TypeScript namespaces.
- Preserve concrete TS namespaces.

## New Features
- Introducing [Verbatim Module Syntax](https://www.typescriptlang.org/tsconfig#verbatimModuleSyntax).
- Enum value will now be inlined whenever possible within a single module, optimizing runtime performance.
- Constant enums will be automatically eliminated when feasible, This can reduce bundle size.
- Added support for cross-referencing enum values.

## Deprecated
`TsEnumConfig` is deprecated
  - The `treat_const_enum_as_enum` transform option is deprecated.
  - The `ts_enum_is_readonly` assumption option is deprecated.

**BREAKING CHANGE:**

TypeScript Config is changed.

**Related issue:**

 - Closes #5197
 - Closes #5259 
 - Closes #7177
 - Closes #7453
 - Closes #7676 
 - Closes #7681
 - Closes #7791 
 - Closes #7961

### CODE REVIEW

**Comment by kdy1** [comment]
Making a visitor multi-pass may make it significantly slower

**Comment by magic-akari** [comment]
> Making a visitor multi-pass may make it significantly slower

I will run a performance benchmark after I finish this pull request.

File: crates/swc_ecma_transforms_typescript/src/typescript.rs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+pub fn typescript(config: Config) -> impl Fold + VisitMut {
+    as_folder(TypeScript { config })
+}
+
+pub(crate) struct TypeScript {
+    pub config: Config,
+}
+
+impl VisitMut for TypeScript {
+    // TODO: program will be skipped in test_fixture
```
Comment (magic-akari) [inline crates/swc_ecma_transforms_typescript/src/typescript.rs:47]: `visit_mut_program` is skipped in  `test_fixture`. Is this expected behavior? @kdy1

Comment (kdy1) [reply crates/swc_ecma_transforms_typescript/src/typescript.rs:47]: No, can you fix it too?

**Comment by magic-akari** [comment]
review request

**Comment by magic-akari** [comment]
# branch typescript
```
test es/full/codegen/es3 ... bench:       19076 ns/iter (+/- 67)

test es/full/codegen/es5 ... bench:       19022 ns/iter (+/- 32)

test es/full/codegen/es2015 ... bench:       19061 ns/iter (+/- 30)

test es/full/codegen/es2016 ... bench:       19085 ns/iter (+/- 37)

test es/full/codegen/es2017 ... bench:       19066 ns/iter (+/- 33)

test es/full/codegen/es2018 ... bench:       19023 ns/iter (+/- 59)

test es/full/codegen/es2019 ... bench:       19050 ns/iter (+/- 32)

test es/full/codegen/es2020 ... bench:       19065 ns/iter (+/- 50)

test es/full/all/es3 ... bench:    87371719 ns/iter (+/- 281454)

test es/full/all/es5 ... bench:    83584655 ns/iter (+/- 663850)

test es/full/all/es2015 ... bench:    64777174 ns/iter (+/- 369595)

test es/full/all/es2016 ... bench:    64382368 ns/iter (+/- 365596)

test es/full/all/es2017 ... bench:    64091412 ns/iter (+/- 137335)

test es/full/all/es2018 ... bench:    63352586 ns/iter (+/- 324142)

test es/full/all/es2019 ... bench:    63037412 ns/iter (+/- 234525)

test es/full/all/es2020 ... bench:    60795876 ns/iter (+/- 361287)

test es/full/parser ... bench:      298452 ns/iter (+/- 243)

test es/full/base/fixer ... bench:        7128 ns/iter (+/- 13)

test es/full/base/resolver_and_hygiene ... bench:       29270 ns/iter (+/- 23)


```

File: crates/swc/tests/tsc-references/ClassAndModuleThatMergeWithStaticFunctionAndExportedFunctionThatShareAName.1.normal.js [hunk 2]
```diff
@@ -43,9 +43,10 @@ var A;
     }();
     A.Point = Point;
     (function(Point) {
-        function Origin() {
+        var Origin = function Origin() {
```
Comment (magic-akari) [inline crates/swc/tests/tsc-references/ClassAndModuleThatMergeWithStaticFunctionAndExportedFunctionThatShareAName.1.normal.js:46]: This change is not my intention.
It's should be function declaration instead of variable declaration.
It works as expected in TypeScript test case.
I have no idea about which pass changed it.

File: crates/swc/tests/tsc-references/exportAssignmentMergedInterface.1.normal.js [hunk 3]
```diff
@@ -8,9 +8,8 @@ define([
 //// [foo_1.ts]
 define([
     "require",
-    "exports",
-    "./foo_0"
-], function(require, exports, _foo_0) {
```
Comment (magic-akari) [inline crates/swc/tests/tsc-references/exportAssignmentMergedInterface.1.normal.js:13]: This is a bugfix.
foo_0 is not used and removed.

```ts repro
// @module: amd
// @filename: foo_0.ts
interface Foo {
	(a: string): void;
	b: string;
}
interface Foo {
	(a: number): number;
	c: boolean;
	d: {x: number; y: number};
}
export = Foo;

// @filename: foo_1.ts
import foo = require("./foo_0");
var x: foo;
x("test");
x(42);
var y: string = x.b;
if(!!x.c){ }
var z = {x: 1, y: 2};
z = x.d;
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?module=2&ts=5.2.2#code/PTAEAEFsHsBMFcA2BTAXKAhpWAoEEAzASxQDss1QDpoB9ABgDoAXAZxyNOeQCcCMAxslAAxGqADeOAJAAKDOlbMenAOYBKdADdoRWAG4ZAI0XK1hgL4cuvfkNHipchaFLxIR3ptfvPPQ9IC6EY0KBikAbDoEgAe6G4evPqgAJ7xvrwWljjIMQAO0DzMoAC8DtCGeGDgxGQU6NR0AIws7ESQBUVU4mU8yACO8ER9sgBEjMCNDKPqhloYPKBx3RU4MWPcSjOG6wAsAEyzOPOLaaBKKqSqpUuMRoZEBLIAhM8xjALqEqBWJ6AAXjdYugmgAaVLofZZHCAsrvAxAA)

**Comment by magic-akari** [comment]
@kdy1 This pull request is ready for review.

File: crates/swc/tests/tsc-references/ClassAndModuleThatMergeWithStaticFunctionAndExportedFunctionThatShareAName.1.normal.js (hunk 2 above)
Comment (kdy1) [reply crates/swc/tests/tsc-references/ClassAndModuleThatMergeWithStaticFunctionAndExportedFunctionThatShareAName.1.normal.js:46]: It looks like https://github.com/swc-project/swc/issues/2793

File: crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js [hunk 4]
```diff
@@ -22,7 +22,15 @@ var A = /*#__PURE__*/ function() {
 var M;
 (function(M) {
     var n;
-    M.n = n;
+    Object.defineProperty(M, "n", {
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js:25]: Is this change intentional?

Comment (kdy1) [reply crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js:25]: It looks like a regression of the output size

Comment (magic-akari) [reply crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js:25]: Yes, `export let/var` in TypeScript means mutable exported bindings.

Comment (magic-akari) [reply crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js:25]: This change is a bugfix of #5259.
If output size is more important, I can change it to have the same output as `export const`.

Comment (kdy1) [reply crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js:25]: I think it's better to optimize for the output size because `isolatedModules: true` is our assumption, and `tsc` works similar under same assumptions

Comment (magic-akari) [reply crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js:25]: OK, I will have a anoter try.

Comment (magic-akari) [reply crates/swc/tests/tsc-references/bitwiseNotOperatorWithStringType.1.normal.js]: @kdy1 done.
I will try to fix #5259 in another pull request.

File: crates/swc/tests/tsc-references/defaultExportsCannotMerge01.2.minified.js [hunk 5]
```diff
@@ -1,24 +1,17 @@
...
-        enumerable: !0,
-        get: all[name]
-    });
-}(exports, {
-    Decl: function() {
-        return Decl;
-    },
-    default: function() {
-        return Decl;
+}), Object.defineProperty(exports, "default", {
```
Comment (magic-akari) [inline crates/swc/tests/tsc-references/defaultExportsCannotMerge01.2.minified.js:9]: I thougnt this was a bug.
But I found we get the same result of tsc now.

https://www.typescriptlang.org/play?target=1&module=1#code/PTAEAEFsHsBMFcA2BTAXKAxtSMB2ArAZwCgQIAXAQwCcBzZc9AUQGUBWY45ADwAdpq5ULGQAzSkiGj4uDOQCW0XKAAiyDIgAUASlABvYqCOhqDeNWUAGANzEAvpx79BoebnLJq4jMlXrE+obGvACM6LjwkABGnrbGoLwATOGRMdS2Dlx8AkK4lJDIhLyUPn4agfFOOaAAbjSg3KAAvKAhNkFGVS511KAAns2gie0dru6e3r4AkhXGDg5AA

File: crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js [hunk 6]
```diff
@@ -2,20 +2,20 @@ var A;
 (function(A) {
     const a = 1;
     const b = 2;
-    ({ a: A.A  } = {
+    const { a: A1 } = { a: A.A } = {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js:5]: Is this correct?

Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js:5]: I think it's OK.
The exported name(`A`) is correct.
The local ident is rename to `A1` because of hygiene/resolved pass.

Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js:5]: I have tested the following input:

```TypeScript
module A {
    const a = 1;
    export const { a: A } = { a };

    console.log(A);
}
```

Output:
```TypeScript
var A;
(function(A) {
    const a = 1;
    const { a: A1 } = { a: A.A } = {
        a
    };
    console.log(A1);
})(A || (A = {}));

```

Comment (kdy1) [reply crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js:5]: I mean, it seems like there's duplicate bindings.

Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js:5]: No, there is only one binding.

```
const { a: A1 } = { a: A.A } = {
            |          ^^^---------- assign `a` to `A.A`
            |-- local binding
```

Comment (magic-akari) [reply crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js:5]: https://github.com/swc-project/swc/blob/06fec63ceb9ff09fb540fa08e5bf9f7816ddbd13/crates/swc_ecma_transforms_typescript/src/transform.rs#L594-L611

**Comment by kdy1** [comment]
image

😄

File: crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js (hunk 6 above)
Comment (kdy1) [reply crates/swc_ecma_transforms_typescript/tests/fixture/issue-3073/4/output.js:5]: Thanks!

**Review by kdy1 — APPROVED** [review]
Thank you so much!!!


---

swc-bump:
 - swc_ecma_codegen
 - swc_ecma_transforms_base
